### PR TITLE
Add a script to run operator locally

### LINKED ISSUE #197: Script and document how to start the custom operator locally
The information is already in `readme`, but it spreads several section and contains some shell magic

### PR DESCRIPTION
**Pull request by zerg-junior**
Closes #197 

This is the skeleton of the script. Once it gets reviewed, I will add the capability to run  locally  a self-build operator image and otherwise polish the code.

### CODE REVIEW

File: run_locally.sh [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+echo "==== CLEAN UP PREVIOUS RUN ==== "
+
+status=$(minikube status --format "{{.MinikubeStatus}}")
+if [ "$status" = "Running" ]; then 
+    minikube delete
+fi
+
+# the kubectl process does the port-forwarding between operator and local ports
+# we restart the process to be able to bind to the same port again (see end of script)
+echo "Kill kubectl process to re-init port-forwarding for minikube"
```
Comment (zerg-junior) [inline run_locally.sh]: check with `ps` that the correct `kubectl` is killed

File: run_locally.sh [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+# 1> redirects stdout to remove the info message about forwarded ports; the message sometimes garbles the cli prompt
+kubectl port-forward "$operatorPod" "$localPort":"$operatorPort" 1> /dev/null &
+
+
+echo "==== RUN HEALTH CHECK ==== "
+checkCmd="curl -L http://127.0.0.1:$localPort/clusters &> /dev/null"
+checkMsg="Wait for port forwarding to take effect"
+retry "$checkCmd"  "$checkMsg"
+
+if [ $? -eq 0 ]; then
```
Comment (zerg-junior) [inline run_locally.sh]: substitute with a direct call to retry

File: run_locally.sh [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+function retry(){
+    cmd="$1"
+    retryMsg="$2"
+    for i in {1..90}; do
+	eval "$cmd"
+	if [ $? -eq 0 ]; then
+         return 0
+        fi
+	echo "$retryMsg"
+        sleep 2
```
Comment (zerg-junior) [inline run_locally.sh]: 1. make indentation consistent. 
2. remove trailing spaces and double space

File: run_locally.sh [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    done
+    
+    return 1
+}
+
+
+echo "==== CLEAN UP PREVIOUS RUN ==== "
+
+status=$(minikube status --format "{{.MinikubeStatus}}")
+if [ "$status" = "Running" ]; then
```
Comment (zerg-junior) [inline run_locally.sh]: what if `minikube stop` ?

**Comment by zerg-junior** [comment]
Thanks to CyberDem0n for the review; I just typed the comments.

File: run_operator_locally.sh [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    echo "==== FORWARD OPERATOR PORT $OPERATOR_PORT TO LOCAL PORT $LOCAL_PORT  ===="
+    
+    local operator_pod
+    operator_pod=$(kubectl get pod -l name=postgres-operator -o jsonpath={.items..metadata.name})
+    
+    # runs in the background to keep current terminal responsive
+    # stdout redirect removes the info message about forwarded ports; the message sometimes garbles the cli prompt
+    kubectl port-forward "$operator_pod" "$LOCAL_PORT":"$OPERATOR_PORT" &> /dev/null &
+    
+    pgrep --newest "kubectl" > "$PATH_TO_PORT_FORWARED_KUBECTL_PID"
```
Comment (CyberDem0n) [inline run_operator_locally.sh]: `$!` -- contains the pid of the last background process

File: run_operator_locally.sh [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+readonly PATH_TO_PORT_FORWARED_KUBECTL_PID="/tmp/kubectl-port-forward.pid"
+readonly LOCAL_PORT="8080"
+readonly OPERATOR_PORT="8080"
+
+
+# minikube needs time to create resources,
+# so the script retries actions until all the resources become available
+function retry(){
+
+    # errexit may break "eval $retry_cmd", so we disable it temporarily
```
Comment (CyberDem0n) [inline run_operator_locally.sh]: Really? Even when you are doing `if  eval "..."; then`?

File: run_operator_locally.sh [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        sleep 3
+    done
+
+    >2& echo "The command $retry_cmd timed out"
+    return 1
+}
+
+
+function display_help(){
+    echo "Usage: ./run_locally.sh [ -r | --rebuild-operator ] [ -h | --help ]"
```
Comment (CyberDem0n) [inline run_operator_locally.sh]: `./run_locally.sh` doesn't match with the script name. Why not to use `"$0"`?

File: run_operator_locally.sh [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    if [[ -e "$PATH_TO_LOCAL_OPERATOR_MANIFEST" ]]; then
+        rm --verbose "$PATH_TO_LOCAL_OPERATOR_MANIFEST"
+    fi
+
+    # the kubectl process does the port-forwarding between operator and local ports
+    # we restart the process to bind to the same port again (see end of script)
+    if [[ -e "$PATH_TO_PORT_FORWARED_KUBECTL_PID" ]]; then
+
+        local pid
+        pid=$(cat "$PATH_TO_PORT_FORWARED_KUBECTL_PID")
```
Comment (CyberDem0n) [inline run_operator_locally.sh]: local pid=$(<$PATH_TO_PORT_FORWARED_KUBECTL_PID)

File: run_operator_locally.sh [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+        # the process dies if a minikube stops between two invocations of the script
+        if ps --pid "$pid" > /dev/null  2>&1; then
```
Comment (CyberDem0n) [inline run_operator_locally.sh]: you don't really need to check that process is alive to kill it:
```bash
if kill $pid; then
    echo "Kill the kubectl..."
fi
```

File: run_operator_locally.sh [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+            echo "Kill the kubectl process responsible for port forwarding for minikube so that we can re-use the same ports for forwarding later..."
+            kill "$pid"
+        fi
+        rm --verbose /tmp/kubectl-port-forward.pid
```
Comment (CyberDem0n) [inline run_operator_locally.sh]: rm $PATH_TO_PORT_FORWARED_KUBECTL_PID

File: run_operator_locally.sh [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    echo "==== FORWARD OPERATOR PORT $OPERATOR_PORT TO LOCAL PORT $LOCAL_PORT  ===="
+
+    local operator_pod
+    operator_pod=$(kubectl get pod -l name=postgres-operator -o jsonpath={.items..metadata.name})
+
+    # runs in the background to keep current terminal responsive
+    # stdout redirect removes the info message about forwarded ports; the message sometimes garbles the cli prompt
+    kubectl port-forward "$operator_pod" "$LOCAL_PORT":"$OPERATOR_PORT" &> /dev/null &
+
+    pgrep --newest "kubectl" > "$PATH_TO_PORT_FORWARED_KUBECTL_PID"
```
Comment (CyberDem0n) [inline run_operator_locally.sh]: echo $! > $PATH_TO_PORT_FORWARED_KUBECTL_PID

File: run_operator_locally.sh [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+    if ! [[ $(basename "$PWD") == "postgres-operator" ]]; then
+        echo "Please execute the script only from the root directory of the Postgres opepator repo."
+        exit 1
+    fi
+
+    trap "echo 'If you observe issues with minikube VM not starting/not proceeding, consider deleting the .minikube dir and/or rebooting before re-running the script'" EXIT
+
+    local should_build_custom_operator=false # used in start_operator()
+    while true
```
Comment (CyberDem0n) [inline run_operator_locally.sh]: Do you really need a loop to check only the first command line argument?

File: run_operator_locally.sh [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+
+    pgrep --newest "kubectl" > "$PATH_TO_PORT_FORWARED_KUBECTL_PID"
+}
+
+
+function check_health(){
+
+    echo "==== RUN HEALTH CHECK ==== "
+
+    local -r check_cmd="curl --location --silent http://127.0.0.1:$LOCAL_PORT/clusters &> /dev/null"
```
Comment (CyberDem0n) [inline run_operator_locally.sh]: curl -sLo /dev/null http://127.0.0.1:$LOCAL_PORT/clusters

**Comment by zerg-junior** [comment]
@CyberDem0n updated

**Comment by zerg-junior** [comment]
:+1:

**Comment by mgomezch** [comment]
:+1:

**Pull request merged**